savant/utils/modules_cache.py

```diff
@@ -0,0 +1,123 @@
+"""Inotify manager module."""
+import logging
+import sys
+from collections import defaultdict
+from importlib import reload
+from importlib.machinery import ModuleSpec
+from types import ModuleType
+from typing import Dict, Optional
+
+from inotify_simple import INotify, flags
+
+from savant.utils.logging import get_logger
+from savant.utils.singleton import SingletonMeta
+from importlib import util as importlib_util
+
+logger = get_logger(__name__)
+
+
+def import_module(spec: ModuleSpec) -> ModuleType:
+    """Instantiate a module from module spec."""
+    module_instance = importlib_util.module_from_spec(spec)
+    spec.loader.exec_module(module_instance)
+    # reloading requires module being put into the modules dict
+    # only do so after the loader.exec_module call as it can fail
+    sys.modules[spec.name] = module_instance
+    return module_instance
+
+
+class ModulesCache(metaclass=SingletonMeta):
+    """Singleton facilitates caching and reloading of user updates
+    PyFunc modules in dev mode.
+    """
+
+    def __init__(self) -> None:
+        self.logger = get_logger(__name__)
+        self.inotify = INotify()
+        self.watch_flags = flags.MODIFY
+
+        self.module_file_to_watch = {}
+        self.watch_to_module_file: Dict[str, str] = {}
+        self.changed_modules = defaultdict(bool)
+        self._cache_module: Dict[str, ModuleType] = {}
+
+    def get_module_instance(self, module_spec: ModuleSpec) -> Optional[ModuleType]:
+        """Get module instance by module specification.
+
+        :param module_spec: user module specification.
+        :return: module instance.
+        """
+        module_file = module_spec.origin
+        if module_file not in self._cache_module:
+            logger.debug(f'{module_file} is not cached, importing.')
+            if module_spec is None:
```

Comment:
Module spec is required, the condition doesn't work here.

Reply:
updated